kernel/src/hil/time.rs

```diff
@@ -1,35 +1,186 @@
+  /// After a successful call to `start`, `is_running` MUST return true.    
+  fn start(&self) -> ReturnCode;
+
+  /// Stops the free-running hardware counter. Valid `ReturnCode` values are:
+  ///   - `ReturnCode::SUCCESS`: the counter is now stopped.
+  ///   - `ReturnCode::EBUSY`: the counter is in use in a way that means it
+  ///   cannot be stopped and is busy.
+  ///   - `ReturnCode::FAIL`: unidentified failure, counter is running.
+  /// After a successful call to `stop`, `is_running` MUST return false.        
+  fn stop(&self) -> ReturnCode;
+
+  /// Resets the counter to 0. If the counter is currently running, this
+  /// may require stopping and restarting it. Valid `ReturnCode` values are:
+  ///    - `ReturnCode::SUCCESS`: the counter was reset to 0.
+  ///    - `ReturnCode::FAIL`: the counter was not reset to 0.    
+  fn reset(&self);
+
+  /// Returns whether the counter is currently running.
+  fn is_running(&self) -> bool;
+
+}
+
+/// Callback handler for when an Alarm fires (a `Counter` reaches a specific
+/// value).
+pub trait AlarmClient: OverflowClient {
+  /// Callback indicating the alarm time has been reached. The alarm
+  /// MUST be disabled when this is called. If a new alarm is needed,
+  /// the client can call `Alarm::set_alarm`.
+  fn alarm(&self);
+}
+
+/// Interface for receiving notification when a particular time
+/// (`Counter` value) is reached. Clients use the
+/// [`AlarmClient`](trait.AlarmClient.html) trait to signal when the
+/// counter has reached a pre-specified value set in
+/// [`set_alarm`](#tymethod.set_alarm). Alarms are intended for
+/// low-level time needs that require precision (i.e., firing on a
+/// precise clock tick). Software that needs more functionality
+/// but can tolerate some jitter should use the `Timer` trait
+/// instead.
+
+pub trait Alarm<'a>: Time {
+
+  /// Specify the callback for when the counter reaches the alarm
+  /// value. If there was a previously installed callback this call
+  /// replaces it.
+  fn set_client(&'a self, client: &'a dyn AlarmClient);    
+
+    
+  /// Specify when the callback should be called and enable it. Tthe
+  /// will be triggered when `Time::now() == now + dt`. The callback
+  /// itself may not run exactly at this time, due to delays. However,
+  /// it it assured to execute *after* `now + dt`: it can be delayed
+  /// but will never fire early. The method takes `now` and `dt`
+  /// rather than a single value denoting the counter value so it can
+  /// distinguish between alarms which have very recently already
+  /// passed and those in the far far future (see #1651).
+  fn set_alarm(&self, now: Self::Ticks, dt: Self::Ticks);
+
+  /// Return the current alarm value. This is undefined at boot and
+  /// otherwise returns `now + dt` from the last call to `set_alarm`.
+  fn get_alarm(&self) -> Self::Ticks;
+
+  /// Disable the alarm and stop it from firing in the future.    
+  fn disarm(&self) -> ReturnCode;
```

Comment:
Shouldn't this have the same note as below about this possibly not clearing a pending callback?

Reply:
It should clear the pending callback. The behavior for `is_armed` is for when the callback is pending but has not executed (e.g., kernel is in a different bottom half that ran first). However, if you call `disarm()` you would make it no longer pending.